### PR TITLE
Bump Kubernetes deps, buf, and provider-aws

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

Closes https://github.com/crossplane/function-sdk-go/pull/125
Closes https://github.com/crossplane/function-sdk-go/pull/134
Closes https://github.com/crossplane/function-sdk-go/pull/175

These all had quite a lot of interactions so I had to do them together.

Note we can't bump to Kubernetes v0.31.0 because it's newer than what the latest crossplane-runtime release is using. We'll need to update crossplane-runtime first.

I'm seriously questioning keeping the dependency on provider-aws. It's massive and has a lot of painful dependencies to fix when we update the import. On the other hand, I experimented with e.g. using provider-nop or provider-kubernetes for examples and didn't feel they provided as clear an example as using an S3 bucket. They also didn't improve as much on the dependency pains as I'd hoped.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -5,94 +5,111 @@ go 1.23
...
 	github.com/go-logr/zapr v1.3.0
 	github.com/google/go-cmp v0.6.0
 	github.com/pkg/errors v0.9.1
-	github.com/upbound/provider-aws v0.47.4
+	github.com/upbound/provider-aws v1.13.1
 	go.uber.org/zap v1.27.0
 	google.golang.org/grpc v1.66.2
 	google.golang.org/grpc/cmd/protoc-gen-go-grpc v1.5.1
-	google.golang.org/protobuf v1.34.2
+	google.golang.org/protobuf v1.34.3-0.20240816073751-94ecbc261689
```
Comment (phisco) [inline go.mod:19]: Any particular reason not to bump it just to 1.34.3?

File: go.mod [hunk 2]
```diff
@@ -5,94 +5,111 @@ go 1.23
...
 	k8s.io/apimachinery v0.30.0
 	k8s.io/utils v0.0.0-20240902221715-702e33fdd3c3
 	sigs.k8s.io/yaml v1.4.0
 )
 
 require (
-	buf.build/gen/go/bufbuild/protovalidate/protocolbuffers/go v1.33.0-20240221180331-f05a6f4403ce.1 // indirect
-	connectrpc.com/connect v1.15.0 // indirect
-	connectrpc.com/otelconnect v0.7.0 // indirect
+	buf.build/gen/go/bufbuild/bufplugin/protocolbuffers/go v1.34.2-20240904181154-a0be11449112.2 // indirect
```
Comment (phisco) [inline go.mod:27]: Nit: We should probably do the same we are doing for Crossplane snd just run it without importing it

File: go.mod (hunk 1 above)
Comment (negz) [reply go.mod:19]: Not sure - this is just what the Go dependency solver decided on. 🤷

File: go.mod [hunk 3]
```diff
@@ -5,94 +5,111 @@ go 1.23
...
-	github.com/bufbuild/buf v1.30.0
+	github.com/bufbuild/buf v1.41.0
 	github.com/crossplane/crossplane-runtime v1.17.0
 	github.com/go-json-experiment/json v0.0.0-20240815175050-ebd3a8989ca1
 	github.com/go-logr/logr v1.4.2
 	github.com/go-logr/zapr v1.3.0
 	github.com/google/go-cmp v0.6.0
 	github.com/pkg/errors v0.9.1
-	github.com/upbound/provider-aws v0.47.4
+	github.com/upbound/provider-aws v1.13.1
```
[the remark this answers is not in the document]
Comment (negz) [reply go.mod:15]: True. Might do that next time.

File: go.mod (hunk 2 above)
Comment (phisco) [reply go.mod:27]: The issue is not whether it works, but the fact that they pull in a bunch of dependency that often conflict with stuff we care about, like Kubernetes dependencies

File: go.mod (hunk 1 above)
Comment (negz) [reply go.mod:19]: Apparently it doesn't exist yet:

```
go: google.golang.org/protobuf@v1.34.3: invalid version: unknown revision v1.34.3
```

**Review by phisco — APPROVED** [review]

**Review by blakeromano — APPROVED** [review]

**Pull request merged**